README.md
# three.interaction

[![npm](https://img.shields.io/npm/v/three.interaction.svg?style=flat-square)](https://github.com/jasonChen1982/three.interaction.js)
[![javascript style guide](https://img.shields.io/badge/code_style-google-brightgreen.svg)](https://google.github.io/styleguide/jsguide.html)
[![Standard Version](https://img.shields.io/badge/release-standard%20version-brightgreen.svg)](https://github.com/conventional-changelog/standard-version)

a full-interaction event manager, help `three.js` binding interaction event more simple

# install

```sh
npm install -S three.interaction
```

# usage

```javascript
import { Scene, PerspectiveCamera, WebGLRenderer, Mesh, BoxGeometry, MeshBasicMaterial } from 'three';
import { Interaction } from 'three.interaction';

const renderer = new WebGLRenderer({ canvas: canvasElement });
const scene = new Scene();
const camera = new PerspectiveCamera(60, width / height, 0.1, 100);

// new a interaction, then you can add interaction-event with your free style
const interaction = new Interaction(renderer, scene, camera);

const cube = new Mesh(
  new BoxGeometry(1, 1, 1),
  new MeshBasicMaterial({ color: 0xffffff }),
);
scene.add(cube);
cube.cursor = 'pointer';
cube.on('click', function(ev) {});
cube.on('touchstart', function(ev) {});
cube.on('touchcancel', function(ev) {});
cube.on('touchmove', function(ev) {});
cube.on('touchend', function(ev) {});
cube.on('mousedown', function(ev) {});
cube.on('mouseout', function(ev) {});
cube.on('mouseover', function(ev) {});
cube.on('mousemove', function(ev) {});
cube.on('mouseup', function(ev) {});
// and so on ...

/**
 * you can also listen on parent-node or any display-tree node,
 * source event will bubble up along with display-tree.
 * you can stop the bubble-up by invoke ev.stopPropagation function.
 */
scene.on('touchstart', ev => {
  console.log(ev);
})
scene.on('touchmove', ev => {
  console.log(ev);
})

```

## Documentation
[documentation][documentation]

## Examples
[examples cube][examples]

[examples cube overlap][examples-overlap]


[documentation]:https://jasonchen1982.github.io/three.interaction.js/docs/ "three.interaction documention page"
[examples]:https://jasonchen1982.github.io/three.interaction.js/examples/interaction/ "three.interaction examples page"
[examples-overlap]:https://jasonchen1982.github.io/three.interaction.js/examples/interaction-overlap/ "three.interaction examples page with overlap"


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.3"></a>
## [0.2.3](https://github.com/jasonChen1982/three.interaction.js/compare/v0.2.3-beta.0...v0.2.3) (2020-07-24)



<a name="0.2.3-beta.0"></a>
## [0.2.3-beta.0](https://github.com/jasonChen1982/three.interaction.js/compare/v0.2.2...v0.2.3-beta.0) (2020-07-24)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/jasonChen1982/three.interaction.js/compare/v0.2.1...v0.2.2) (2018-09-06)


### Bug Fixes

* **recursive:** mistake again, should use recursive mode to intersect objects ([5d9c99d](https://github.com/jasonChen1982/three.interaction.js/commit/5d9c99d))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/jasonChen1982/three.interaction.js/compare/v0.2.0...v0.2.1) (2018-09-04)


### Bug Fixes

* **invoke:** use three ray caster with a wrong way, my bad ([4e165a8](https://github.com/jasonChen1982/three.interaction.js/commit/4e165a8))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/jasonChen1982/three.interaction.js/compare/v0.1.0...v0.2.0) (2018-09-03)


### Features

* **intersects:** add intersects for event ([5be6e2b](https://github.com/jasonChen1982/three.interaction.js/commit/5be6e2b))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.9...v0.1.0) (2018-08-24)


### Features

* **layer:** add interaction for layer ([6cd1c5a](https://github.com/jasonChen1982/three.interaction.js/commit/6cd1c5a))



<a name="0.0.9"></a>
## [0.0.9](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.8...v0.0.9) (2018-03-13)


### Bug Fixes

* **error:** fixed event was undefined in onClick function ([08a91f0](https://github.com/jasonChen1982/three.interaction.js/commit/08a91f0))



<a name="0.0.8"></a>
## [0.0.8](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.7...v0.0.8) (2018-03-13)



<a name="0.0.7"></a>
## [0.0.7](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.6...v0.0.7) (2018-01-31)


### Features

* **patch:** remove object3d patch about event emit ([edc6746](https://github.com/jasonChen1982/three.interaction.js/commit/edc6746))
* **peer:** peer dependencies three at 0.89.0 ([93343de](https://github.com/jasonChen1982/three.interaction.js/commit/93343de))
* **utils:** remove some redundant function ([da46a97](https://github.com/jasonChen1982/three.interaction.js/commit/da46a97))



<a name="0.0.6"></a>
## [0.0.6](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.5...v0.0.6) (2017-12-21)


### Features

* **struct:** adjust modules struct ([1a58af2](https://github.com/jasonChen1982/three.interaction.js/commit/1a58af2))



<a name="0.0.5"></a>
## [0.0.5](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.4...v0.0.5) (2017-12-13)


### Bug Fixes

* **touch:** fixed `touchmove` and `touchend` impl like browser ([3a99d11](https://github.com/jasonChen1982/three.interaction.js/commit/3a99d11))



<a name="0.0.4"></a>
## [0.0.4](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.3...v0.0.4) (2017-12-08)


### Bug Fixes

* **addEvents:** fixed multiple bind event, when repeat invoke addEvents ([15acda7](https://github.com/jasonChen1982/three.interaction.js/commit/15acda7))



<a name="0.0.3"></a>
## [0.0.3](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.2...v0.0.3) (2017-12-08)


### Features

* **interaction:** add interaction impl ([8191b1c](https://github.com/jasonChen1982/three.interaction.js/commit/8191b1c))
* **target:** put target into interactionData ([82aed3d](https://github.com/jasonChen1982/three.interaction.js/commit/82aed3d))



<a name="0.0.2"></a>
## [0.0.2](https://github.com/jasonChen1982/three.interaction.js/compare/v0.0.1...v0.0.2) (2017-12-06)


### Bug Fixes

* **move:** move event was not same with browser ([05991dc](https://github.com/jasonChen1982/three.interaction.js/commit/05991dc))


### Features

* **click:** add really click event ([b1abb5a](https://github.com/jasonChen1982/three.interaction.js/commit/b1abb5a))
* **script:** adjust release script ([e42b357](https://github.com/jasonChen1982/three.interaction.js/commit/e42b357))



<a name="0.0.1"></a>
## 0.0.1 (2017-12-05)


### Features

* **init:** finish interaction ([cbb9697](https://github.com/jasonChen1982/three.interaction.js/commit/cbb9697))
* **readme:** add some readme ([363eda1](https://github.com/jasonChen1982/three.interaction.js/commit/363eda1))


src/interaction/Interaction.js
import Ticker from '../utils/Ticker';
import InteractionManager from './InteractionManager';

/**
 * The interaction manager deals with mouse, touch and pointer events. Any DisplayObject can be interactive
 * if its interactive parameter is set to true
 * This manager also supports multitouch.
 *
 * reference to [pixi.js](http://www.pixijs.com/) impl
 *
 * @example
 * import { Scene, PerspectiveCamera, WebGLRenderer, Mesh, BoxGeometry, MeshBasicMaterial } from 'three';
 * import { Interaction } from 'three.interaction';
 * const renderer = new WebGLRenderer({ canvas: canvasElement });
 * const scene = new Scene();
 * const camera = new PerspectiveCamera(60, width / height, 0.1, 100);
 *
 * const interaction = new Interaction(renderer, scene, camera);
 * // then you can bind every interaction event with any mesh which you had `add` into `scene` before
 * const cube = new Mesh(
 *   new BoxGeometry(1, 1, 1),
 *   new MeshBasicMaterial({ color: 0xffffff }),
 * );
 * scene.add(cube);
 * cube.on('touchstart', ev => {
 *   console.log(ev);
 * });
 *
 * cube.on('mousedown', ev => {
 *   console.log(ev);
 * });
 *
 * cube.on('pointerdown', ev => {
 *   console.log(ev);
 * });
 * // and so on ...
 *
 * // you can also listen on parent-node or any display-tree node,
 * // source event will bubble up along with display-tree.
 * // you can stop the bubble-up by invoke ev.stopPropagation function.
 * scene.on('touchstart', ev => {
 *   console.log(ev);
 * })
 *
 * @class
 * @extends InteractionManager
 */
class Interaction extends InteractionManager {
  /**
   * @param {WebGLRenderer} renderer - A reference to the current renderer
   * @param {Scene} scene - A reference to the current scene
   * @param {Camera} camera - A reference to the current camera
   * @param {Object} [options] - The options for the manager.
   * @param {Boolean} [options.autoPreventDefault=false] - Should the manager automatically prevent default browser actions.
   * @param {Boolean} [options.autoAttach=false] - Should the manager automatically attach target element.
   * @param {Number} [options.interactionFrequency=10] - Frequency increases the interaction events will be checked.
   */
  constructor(renderer, scene, camera, options) {
    options = Object.assign({ autoAttach: false }, options);
    super(renderer, scene, camera, options);

    /**
     * a ticker
     *
     * @private
     * @member {Ticker}
     */
    this.ticker = new Ticker();

    /**
     * update for some over event
     *
     * @private
     */
    this.update = this.update.bind(this);

    this.on('addevents', () => {
      this.ticker.on('tick', this.update);
    });

    this.on('removeevents', () => {
      this.ticker.off('tick', this.update);
    });

    this.setTargetElement(this.renderer.domElement);
  }
}

export default Interaction;
